larray/core/array.py

```diff
@@ -5572,6 +5542,201 @@ def extend(self, axis, other):
+         a1   3    44    45   4   5
+        """
+
+        # XXX: unsure we should have arr1.insert(arr3, before='b1,b2') result in (see unit tests):
+
+        # a\\b  b0  b0.1  b1  b0.2  b2
+        #  a0   0    42   1    43   2
+        #  a1   3    44   4    45   5
+
+        # we might to implement the following instead:
+
+        # a\\b  b0  b0.1  b0.2  b1  b0.1  b0.2  b2
+        #  a0   0    42    43   1    42    43   2
+        #  a1   3    44    45   4    44    45   5
+
+        # The later looks less useful and could be emulated easily via:
+        # arr1.insert([arr3, arr3], before='b1,b2')
+        # while the above is a bit harder to achieve manually:
+        # arr1.insert([arr3[[b]] for b in arr3.b], before=['b1', 'b2'])
+        # but the later is *probably* more intuitive (and wouldn't suffer from the inefficiency we currently have).
+
+        # XXX: when we have several lists, we implicitly match them by position, which we should avoid for the usual
+        # reason, but I am unsure what the best syntax for that would be.
+
+        # the goal is to get this result
+
+        # a\b  b0  b0.5  b1  b1.5  b2
+        #  a0   0     8   1     9   2
+        #  a1   3     8   4     9   5
+
+        # When the inserted arrays already contain a label, this seems reasonably readable:
+
+        # >>> arr1 = ndtest((2, 3))
+        # >>> arr1
+        # a\\b  b0  b1  b2
+        #  a0   0   1   2
+        #  a1   3   4   5
+        # >>> arr2 = full('b=b0.5', 8)
+        # >>> arr2
+        # b  b0.5
+        #       8
+        # >>> arr3 = full('b=b1.5', 9)
+        # >>> arr3
+        # b  b1.5
+        #       9
+        # >>> arr1.insert(before={'b1': arr2, 'b2': arr3})
+        # a\\b  b0  b0.5  b1  b1.5  b2
+        #  a0   0     8   1     9   2
+        #  a1   3     8   4     9   5
+
+        # When the inserted arrays/values have no label, this does not really convince me and it prevents using after
+        # or pos.
+
+        # >>> arr1.insert(value={'b0.5': ('b1', 8), 'b1.5': ('b2', 9)})
+        # a\b  b0  b0.5  b1  b1.5  b2
+        #  a0   0     8   1     9   2
+        #  a1   3     8   4     9   5
+
+        # This works with both after and pos and we could support it along with the above syntax when no label is
+        # needed. Problem: label, value is arbitrary and as such potentially hard to remember.
+
+        # >>> arr1.insert(before={'b1': ('b0.5', 8), 'b2': ('b1.5', 9)})
+        # a\b  b0  b0.5  b1  b1.5  b2
+        #  a0   0     8   1     9   2
+        #  a1   3     8   4     9   5
+
+        # This is shorter but not readable enough/even more arbitrary than the previous option.
+
+        # >>> arr1.insert([(8, 'b1', 'b0.5'), (9, 'b2', 'b1.5')])
+        # a\b  b0  b0.5  b1  b1.5  b2
+        #  a0   0     8   1     9   2
+        #  a1   3     8   4     9   5
+
+        # This is readable but odd and not much gained (except efficiency) compared with multiple insert calls
```

Comment:
Or a list of dictionary {'value': XX, 'before': YY, 'label': ZZ}. Why not let users to use insert in a loop?

Reply:
they can of course, but it is awfully slow compared to one insert call (because it reallocates & copies the array for each iteration of the loop) while it is possible to do it in one call.

Reply:
The list of dict seems like a good idea. Might just be the compromise syntax I was looking for. I have to see what it looks like in a example before I can decide on this, though.